Documentation/Common-Gotchas.md

```diff
@@ -0,0 +1,232 @@
+
+```swift
+struct User: Decodable {
+  let id: Int
+  let name: String
+  let bio: String
+  let smallAvatar: String
+  let mediumAvatar: String
+  let largeAvatar: String
+
+  static func decode(j: JSON) -> Decoded<User> {
+    return curry(User.init)
+      <^> j <| "id"
+      <*> j <| "name"
+      <*> j <| "bio"
+      <*> j <| "small_avatar"
+      <*> j <| "medium_avatar"
+      <*> j <| "large_avatar"
+  }
+}
+```
+
+With our assumption on `curry`, we would get a Swift compiler error when doing
+`curry(User.init)` because `User.init` takes 6 arguments. Even though the API
+has decided to send us back separate fields for each avatar size, there is a
+very clear normalization we can do in our model:
+
+```swift
+struct User {
+  let id: Int
+  let name: String
+  let bio: String
+  let avatar: Avatar
+
+  struct Avatar {
+    let small: String
+    let medium: String
+    let large: String
+  }
+}
+```
+
+Now `User.init` has 4 arguments and `User.Avatar.init` has 3 arguments, so we
+are within the assumed limitations of `curry`. But, it is also less clear how we
+should write `User.decode`. How can we extract the 3 avatar fields from `JSON`
+to build `User.Avatar` first, and then use that to build `User`? The key insight
+is to see that `j <| "key"` returns a `Decoded<T>` value, representing one step
+of decoding. So, we can decode a `User.Avatar` first using `j`, and then just
+plug that directly into `curry(User.init)`:
+
+```swift
+extension User.Avatar: Decodable {
+  static func decode(j: JSON) -> Decoded<User.Avatar> {
+    return curry(User.Avatar.init)
+      <^> j <| "small_avatar"
+      <*> j <| "medium_avatar"
+      <*> j <| "large_avatar"
+  }
+}
+
+extension User: Decodable {
+  static func decode(j: JSON) -> Decoded<User> {
+    return curry(User.init)
+      <^> j <| "id"
+      <*> j <| "name"
+      <*> j <| "bio"
+      <*> User.Avatar.decode(j)
+  }
+}
+```
+
+We have now successfully gotten around `curry`’s limitations, but also produced
+a better, normalized model! We also retain all of the great error handling that
+Argo provides, so if there is a decoding error in `User.Avatar` it will
+propogate through to a decoding error in `User`.
+
+## Compilation errors
+
+Swift’s compile errors can be mystifying if you are not intimately familiar with
+the internals of Argo. Here we describe some of the most common errors and how
+to fix them.
+
+### Complex expressions
+
+Considering that the implementation of `decode` on a model consists of fully
+currying a method and composing it with many small pieces, each of which may be
+the composition of pieces, it is not surprising that Swift sometimes has
+difficulty compiling the expression in a reasonable amount of time. This
+manifests itself with the following error:
+
+> Expression was too complex to be solved in reasonable time; consider breaking
+up the expression into distinct sub-expressions.
+
+Fortunately this is easy to solve. A first step is to simply store
```

Comment:
We could drop the word "simply" here: it would make the sentence shorter, and also alienate fewer people.